File: apps/web/src/types/transaction.ts
export type TransactionType = 'expense' | 'income' | 'investment' | 'transfer'
export type TransactionOrigin = 'manual' | 'open_finance' | 'csv_import'

export interface Transaction {
  id: string
  account_id: string | null
  category_id: string | null
  description: string
  amount: number
  type: TransactionType
  date: string
  notes: string | null
  origin: TransactionOrigin
  is_reconciled: boolean
  is_recurring: boolean
  ai_suggested_category_id: string | null
  ai_confidence: number | null
  created_at: string
}

export interface TransactionListResponse {
  items: Transaction[]
  total: number
  page: number
  page_size: number
}

export interface TransactionParseResult {
  description: string
  amount: number
  type: TransactionType
  category_id: string | null
  category_name: string | null
  date: string
  confidence: number
  alternatives: { category_name: string; confidence: number }[]
}

export interface CsvImportResult {
  created: number
  reconciled: number
  skipped: number
}

export type ImportFileFormat = 'csv' | 'xls' | 'xlsx' | 'pdf'
export type ImportDocumentType = 'extrato' | 'fatura'

export interface ImportPreviewTransaction {
  date: string
  description: string
  /** Valor serializado como string decimal pela API (ex.: "123.45"). */
  amount: string
  type: TransactionType
  category_id: string | null
}

/** Resposta de POST /v1/transactions/import/preview. */
export interface ImportPreviewResult {
  file_format: ImportFileFormat
  document_type: ImportDocumentType
  confidence: number
  transactions: ImportPreviewTransaction[]
  warnings: string[]
}

/** Resposta de POST /v1/transactions/import/commit. */
export interface ImportCommitResult {
  created: number
  reconciled: number
  skipped: number
}

File: apps/api/app/schemas/transaction.py
import datetime as dt
import uuid
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

TransactionType = Literal["expense", "income", "investment", "transfer"]
TransactionOrigin = Literal["manual", "open_finance", "csv_import"]


class TransactionCreate(BaseModel):
    description: str = Field(min_length=1, max_length=500)
    amount: Decimal = Field(gt=0)
    type: TransactionType
    date: dt.date
    account_id: uuid.UUID | None = None
    category_id: uuid.UUID | None = None
    notes: str | None = None


class TransactionUpdate(BaseModel):
    description: str | None = Field(default=None, min_length=1, max_length=500)
    amount: Decimal | None = Field(default=None, gt=0)
    type: TransactionType | None = None
    date: dt.date | None = None
    account_id: uuid.UUID | None = None
    category_id: uuid.UUID | None = None
    notes: str | None = None


class TransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID | None
    category_id: uuid.UUID | None
    description: str
    amount: Decimal
    type: str
    date: dt.date
    notes: str | None
    origin: str
    is_reconciled: bool
    is_recurring: bool
    ai_suggested_category_id: uuid.UUID | None
    ai_confidence: float | None
    created_at: dt.datetime


class TransactionListResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    page: int
    page_size: int


class TransactionParseRequest(BaseModel):
    text: str = Field(min_length=1, max_length=500)


class ParsedCategoryAlternative(BaseModel):
    category_name: str
    confidence: float


class TransactionParseResponse(BaseModel):
    description: str
    amount: Decimal
    type: TransactionType
    category_id: uuid.UUID | None = None
    category_name: str | None = None
    date: dt.date
    confidence: float
    alternatives: list[ParsedCategoryAlternative] = []


class CsvImportResult(BaseModel):
    created: int
    reconciled: int
    skipped: int
